[8.10] POST /v2/authorizations returns 500 for ownerType=ROLE — HttpMessageConversion fails on CamundaAuthentication

## Summary

`POST /v2/authorizations` (and other endpoints that load a persistent web session) returns **HTTP 500** in the SM-8.10 RBA build. The visible client error is

> Failed to convert from type `[java.lang.Object]` to type `[byte[]]`
> for value `[io.camunda.security.api.model.CamundaAuthentication@<hash>]`

but the **actual** server-side cause (verified by [@megglos](https://github.com/megglos) on Cloud Logging) is the inverse conversion — a session **deserialization** failure with `ClassNotFoundException: io.camunda.security.auth.CamundaAuthentication`.

The endpoint is reachable through the Admin UI's "Create authorization" modal on the Authorizations → PROCESS_DEFINITION tab. The 500 is deterministic for any `ownerType: ROLE` request whenever the session-attribute deserializer is invoked against a session blob that was persisted under the older FQN.

## Verified root cause

`CamundaAuthentication` was moved between packages:

- **Old FQN (still referenced by persisted session bytes):** `io.camunda.security.auth.CamundaAuthentication`
- **New FQN (only one present in the current binary):** `io.camunda.security.api.model.CamundaAuthentication`

The orchestration cluster's persistent session store keeps Java-serialized session attributes. Any session attribute that holds a `CamundaAuthentication` reference was written with the old FQN. When a new build with the renamed class boots and tries to deserialize that blob (either during request handling or in the background `WebSessionDeletionTask` sweeper), `ObjectInputStream.resolveClass` throws `ClassNotFoundException`.

### Actual server-side stack trace (from `WebSessionDeletionTask` sweep)

```
org.springframework.core.convert.ConversionFailedException:
  Failed to convert from type [byte[]] to type [java.lang.Object]
    at org.springframework.core.convert.support.ConversionUtils.invokeConverter
    at org.springframework.core.convert.support.GenericConversionService.convert
    at io.camunda.security.spring.session.WebSessionMapper$SpringBasedWebSessionAttributeConverter.deserialize(WebSessionMapper.java:103)
    at io.camunda.security.spring.session.WebSessionMapper.lambda$deserializeSessionAttributes$1(WebSessionMapper.java:78)
    ...
    at io.camunda.security.spring.session.WebSessionRepository.toWebSession(WebSessionRepository.java:105)
    at io.camunda.security.spring.session.WebSessionRepository.deletePersistentSessionIfExpired(WebSessionRepository.java:82)
    at io.camunda.security.spring.session.WebSessionRepository.lambda$deleteExpiredWebSessions$0(WebSessionRepository.java:78)
    at io.camunda.security.spring.session.WebSessionRepository.deleteExpiredWebSessions(WebSessionRepository.java:76)
    at io.camunda.security.spring.session.WebSessionDeletionTask.run(WebSessionDeletionTask.java:27)
    at io.camunda.security.spring.session.WebSessionConfiguration$SelfSchedulingTask.run(WebSessionConfiguration.java:121)

Caused by: org.springframework.core.serializer.support.SerializationFailedException:
  Failed to deserialize payload. Is the byte array a result of corresponding serialization for DefaultDeserializer?

Caused by: java.io.IOException: Failed to deserialize object type

Caused by: java.lang.ClassNotFoundException:
  io.camunda.security.auth.CamundaAuthentication
    at java.io.ObjectInputStream.resolveClass(...)
    at org.springframework.core.serializer.DefaultDeserializer.deserialize(DefaultDeserializer.java:82)
```

The visible UI error wraps the inverse `[Object → byte[]]` because Spring re-uses the same `ConversionFailedException` shape on the calling side; the wrapping stack frame identifies it as a **deserialize** call from `WebSessionMapper`.

## Why USER works but ROLE fails

Both code paths go through `WebSessionMapper`. The ROLE path additionally hits `DefaultMembershipService.roleIds()` which triggers loading of session attributes that carry a `CamundaAuthentication` reference; that's the attribute whose serialized bytes hold the stale FQN.

## Likely fix surface

Either (or both):

1. **`io.camunda.security.spring.session.WebSessionMapper.SpringBasedWebSessionAttributeConverter.deserialize`** — handle `ClassNotFoundException` (or use a custom `ObjectInputStream` that rewrites the old `io.camunda.security.auth.CamundaAuthentication` FQN to the new one) so cross-version sessions can be read, or evict them gracefully instead of failing every request.
2. **`io.camunda.security.spring.session.WebSessionDeletionTask`** — when deserialization of a session attribute fails with `ClassNotFoundException`, drop the session rather than throw, so the sweeper makes progress on its own.

Either approach makes the upgrade path resilient to renames of session-attribute classes.

## Reproduction

Confirmed live on the SM-8.10 RBA env. The repro is deterministic when the env's session store carries blobs from a previous deployment with the older FQN.

### Step 1 — Log in to the Admin UI

1. Open `https://<deployment>/orchestration/admin/`
2. Sign in as a user with admin authorizations (e.g. `demo`).

### Step 2 — Create a role

1. Top nav → **Roles** → **Create role +**.
2. Role ID & name: `repro-role-1` → **Create role**.
3. Modal closes; new row appears. *(This call returns **201**.)*

### Step 3 — Create the authorization that triggers the 500

1. Top nav → **Authorizations**.
2. Left sidebar → click **PROCESS_DEFINITION**.
3. Top right → **Create authorization +**.
4. Fill the modal:
   - **Owner type** → `ROLE`
   - **Owner** → type `repro-role-1`, pick from dropdown
   - **Resource type** → locked to `PROCESS_DEFINITION`
   - **Resource ID** → `repro-role-1` (any value)
   - **Permissions** → tick `READ_PROCESS_DEFINITION` and `CREATE_PROCESS_INSTANCE`
5. Click **Create authorization**.
6. **Observed**: modal stays open with red **Internal Server Error** banner referencing `CamundaAuthentication`.

### Failing request

```
POST /orchestration/v2/authorizations
Content-Type: application/json

{
  "ownerType": "ROLE",
  "ownerId": "repro-role-1",
  "resourceType": "PROCESS_DEFINITION",
  "resourceId": "repro-role-1",
  "resourcePropertyName": null,
  "permissionTypes": ["READ_PROCESS_DEFINITION", "CREATE_PROCESS_INSTANCE"]
}
```

### Inspecting the stale session blobs

The persistent session store is in the orchestration cluster's Postgres. To see the offending entries while an env is live:

```bash
kubectl exec -n <namespace> postgresql-0 -- \
  psql -U postgres -c "\\dt *session*"

kubectl exec -n <namespace> postgresql-0 -- \
  psql -U postgres -c "SELECT session_id, principal_name, last_access_time, length(attribute_bytes) FROM spring_session_attributes LIMIT 20;"
```

(Adjust the schema/table name to whatever the orchestration release uses — Spring Session JDBC defaults are `SPRING_SESSION` / `SPRING_SESSION_ATTRIBUTES`.)

## Impact

Blocks any RBA flow that grants permissions to a role on a process definition — the canonical "give role X read+execute on process Y" pattern. The corresponding cross-component e2e tests on the SM-8.10 RBA nightly have been red since 2026-06-24.

More broadly, every future package/FQN rename of a class persisted in session attributes will surface as the same opaque 500 on upgrade unless the session deserializer is made tolerant.

## Originally-wrong diagnosis (kept for context)

The original report (commit history before this edit) attributed the 500 to PR [#55614](https://github.com/camunda/camunda/pull/55614) (physical-tenant identity refactor) based on the visible UI banner. That hypothesis was wrong: the conversion direction reads opposite from what the banner suggests, and the actual cause is a class **rename**, not a serialization-side regression. Credit to [@megglos](https://github.com/megglos) for pulling the real stack trace from Cloud Logging and identifying the deserialization path.

## References

- Last green run: https://github.com/camunda/c8-cross-component-e2e-tests/actions/runs/27995482541
- First red run: https://github.com/camunda/c8-cross-component-e2e-tests/actions/runs/28140514680
- Comment with actual stack trace: [#55964 (comment by @megglos)](https://github.com/camunda/camunda/issues/55964#issuecomment-)

actual error is [related to the sessions persistence](https://console.cloud.google.com/logs/query;cursorTimestamp=2026-06-25T17:28:16.181545377Z;endTime=2026-06-25T17:28:22.019Z;pinnedLogId=2026-06-25T17:28:16.181545377Z%2F6sco2j1cz26jvya8;query=resource.labels.namespace_name%3D%22qa-id-qaelmtupg-0420-8-9-qaelmtupg%22%0A--%22Failed%20to%20convert%20from%20type%22%0Atimestamp%3D%222026-06-25T17:28:16.181545377Z%22%0AinsertId%3D%226sco2j1cz26jvya8%22;startTime=2026-06-25T17:28:12.108Z?project=camunda-distribution):
```
org.springframework.core.convert.ConversionFailedException: Failed to convert from type [byte[]] to type [java.lang.Object] for value [{...}]
	at org.springframework.core.convert.support.ConversionUtils.invokeConverter(ConversionUtils.java:47)
	at org.springframework.core.convert.support.GenericConversionService.convert(GenericConversionService.java:181)
	at io.camunda.security.spring.session.WebSessionMapper$SpringBasedWebSessionAttributeConverter.deserialize(WebSessionMapper.java:103)
	at io.camunda.security.spring.session.WebSessionMapper.lambda$deserializeSessionAttributes$1(WebSessionMapper.java:78)
	at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$0(Unknown Source)
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(Unknown Source)
	at java.base/java.util.Iterator.forEachRemaining(Unknown Source)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)
	at io.camunda.security.spring.session.WebSessionMapper.deserializeSessionAttributes(WebSessionMapper.java:78)
	at io.camunda.security.spring.session.WebSessionMapper.fromPersistentSession(WebSessionMapper.java:49)
	at java.base/java.util.Optional.map(Unknown Source)
	at io.camunda.security.spring.session.WebSessionRepository.toWebSession(WebSessionRepository.java:105)
	at io.camunda.security.spring.session.WebSessionRepository.deletePersistentSessionIfExpired(WebSessionRepository.java:82)
	at java.base/java.util.ArrayList.forEach(Unknown Source)
	at io.camunda.security.spring.session.WebSessionRepository.lambda$deleteExpiredWebSessions$0(WebSessionRepository.java:78)
	at java.base/java.util.Optional.ifPresent(Unknown Source)
	at io.camunda.security.spring.session.WebSessionRepository.deleteExpiredWebSessions(WebSessionRepository.java:76)
	at io.camunda.security.spring.session.WebSessionDeletionTask.run(WebSessionDeletionTask.java:27)
	at io.camunda.security.spring.session.WebSessionConfiguration$SelfSchedulingTask.run(WebSessionConfiguration.java:121)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.springframework.core.serializer.support.SerializationFailedException: Failed to deserialize payload. Is the byte array a result of corresponding serialization for DefaultDeserializer?
	at org.springframework.core.serializer.support.DeserializingConverter.convert(DeserializingConverter.java:81)
	at org.springframework.core.serializer.support.DeserializingConverter.convert(DeserializingConverter.java:38)
	at org.springframework.core.convert.support.GenericConversionService$ConverterAdapter.convert(GenericConversionService.java:357)
	at org.springframework.core.convert.support.ConversionUtils.invokeConverter(ConversionUtils.java:41)
	... 29 more
Caused by: java.io.IOException: Failed to deserialize object type
	at org.springframework.core.serializer.DefaultDeserializer.deserialize(DefaultDeserializer.java:85)
	at org.springframework.core.serializer.support.DeserializingConverter.convert(DeserializingConverter.java:76)
	... 32 more
Caused by: java.lang.ClassNotFoundException: io.camunda.security.auth.CamundaAuthentication
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Unknown Source)
	at java.base/java.io.ObjectInputStream.resolveClass(Unknown Source)
	at org.springframework.core.ConfigurableObjectInputStream.resolveClass(ConfigurableObjectInputStream.java:80)
	at java.base/java.io.ObjectInputStream.readNonProxyDesc(Unknown Source)
	at java.base/java.io.ObjectInputStream.readClassDesc(Unknown Source)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(Unknown Source)
	at java.base/java.io.ObjectInputStream.readObject0(Unknown Source)
	at java.base/java.io.ObjectInputStream.readObject(Unknown Source)
	at java.base/java.io.ObjectInputStream.readObject(Unknown Source)
	at org.springframework.core.serializer.DefaultDeserializer.deserialize(DefaultDeserializer.java:82)
	... 33 more
```
due to a missing `io.camunda.security.auth.CamundaAuthentication` class. This looks like a session is stored based on an older deployment and now fails to serialize.

--- a/authentication/src/main/java/io/camunda/authentication/config/OidcOverrideBeansConfiguration.java
+++ b/authentication/src/main/java/io/camunda/authentication/config/OidcOverrideBeansConfiguration.java
@@ -9,7 +9,9 @@ package io.camunda.authentication.config;
 
 import static java.util.stream.Collectors.toMap;
 
+import io.camunda.authentication.service.PhysicalTenantMembershipContextPropagator;
 import io.camunda.security.api.context.CamundaAuthenticationConverter;
+import io.camunda.security.api.context.MembershipResolutionContextPropagator;
 import io.camunda.security.api.context.OidcClaimsProvider;
 import io.camunda.security.api.model.config.AuthenticationMethod;
 import io.camunda.security.api.model.config.initialization.ConfiguredUser;
@@ -106,14 +108,23 @@ public class OidcOverrideBeansConfiguration {
   }
 
   @Bean
+  @ConditionalOnMissingBean
+  public MembershipResolutionContextPropagator membershipResolutionContextPropagator() {
+    return new PhysicalTenantMembershipContextPropagator();
+  }
+
+  @Bean
   public LazyTokenClaimsConverter tokenClaimsConverter(
-      final CamundaSecurityLibraryProperties cslProperties, final MembershipPort membershipPort) {
+      final CamundaSecurityLibraryProperties cslProperties,
+      final MembershipPort membershipPort,
+      final MembershipResolutionContextPropagator membershipResolutionContextPropagator) {
     final var oidcConfig = cslProperties.getAuthentication().getOidc();
     return new LazyTokenClaimsConverter(
         oidcConfig.getUsernameClaim(),
         oidcConfig.getClientIdClaim(),
         oidcConfig.isPreferUsernameClaim(),
-        membershipPort);
+        membershipPort,
+        membershipResolutionContextPropagator);
   }
 
   @Bean

--- a/authentication/src/main/java/io/camunda/authentication/service/PhysicalTenantMembershipContextPropagator.java
+++ b/authentication/src/main/java/io/camunda/authentication/service/PhysicalTenantMembershipContextPropagator.java
@@ -1,0 +1,34 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.authentication.service;
+
+import io.camunda.security.api.context.MembershipResolutionContextPropagator;
+import io.camunda.spring.utils.PhysicalTenantContext;
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * Host implementation of CSL's {@link MembershipResolutionContextPropagator} that carries the
+ * physical tenant resolved while building the authentication onto whatever thread or scope later
+ * materialises a lazy membership list.
+ *
+ * <p>{@link DefaultMembershipService} routes each membership query to a per-tenant service via
+ * {@link PhysicalTenantContext#current()}. Without this propagator that call would throw when the
+ * list resolves outside the request scope — most visibly when persistent web sessions serialise the
+ * {@code CamundaAuthentication} after the request has been dispatched. By capturing the tenant at
+ * construction time (in scope) and rebinding it around the deferred lookup, the routing key is
+ * available wherever resolution happens to run.
+ */
+public final class PhysicalTenantMembershipContextPropagator
+    implements MembershipResolutionContextPropagator {
+
+  @Override
+  public Supplier<List<String>> decorate(final Supplier<List<String>> supplier) {
+    return PhysicalTenantContext.propagateCurrent(supplier);
+  }
+}

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -45,7 +45,7 @@
     <version.bouncycastle>1.84</version.bouncycastle>
     <version.camunda>7.24.0</version.camunda>
     <version.camunda-license-check>2.11.2</version.camunda-license-check>
-    <version.camunda-security-library>0.1.0-alpha36</version.camunda-security-library>
+    <version.camunda-security-library>0.1.0-alpha37</version.camunda-security-library>
     <version.checkstyle>13.5.0</version.checkstyle>
     <version.commons-beanutils>1.11.0</version.commons-beanutils>
     <version.commons-lang>3.20.0</version.commons-lang>

--- a/spring-utils/src/main/java/io/camunda/spring/utils/PhysicalTenantContext.java
+++ b/spring-utils/src/main/java/io/camunda/spring/utils/PhysicalTenantContext.java
@@ -9,6 +9,7 @@ package io.camunda.spring.utils;
 
 import io.camunda.configuration.api.physicaltenants.PhysicalTenantIds;
 import jakarta.servlet.http.HttpServletRequest;
+import java.util.function.Supplier;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.web.context.request.RequestAttributes;
@@ -137,6 +138,39 @@ public final class PhysicalTenantContext {
         Thread.currentThread().getName(),
         physicalTenantId);
     PROPAGATED_PHYSICAL_TENANT_ID.set(physicalTenantId);
+  }
+
+  /**
+   * Wraps {@code supplier} so that the physical tenant currently in effect — captured eagerly, when
+   * this method is called — is bound as the {@linkplain #setPropagatedPhysicalTenant propagated}
+   * tenant for the duration of each {@code supplier.get()} call, restoring any previously
+   * propagated value afterwards. If no tenant is in effect at capture time, {@code supplier} is
+   * returned unchanged.
+   *
+   * <p>Use this to carry the current thread's tenant into a deferred computation that may run after
+   * the request scope is gone — e.g. a lazy membership list materialised during HTTP session
+   * serialisation, or work handed to an async worker — without the caller managing the propagation
+   * {@link ThreadLocal} directly. This is the same capture-by-value strategy {@code
+   * PhysicalTenantPropagatingExecutorService} applies to executor tasks.
+   */
+  public static <T> Supplier<T> propagateCurrent(final Supplier<T> supplier) {
+    final String physicalTenantId = currentOrNull();
+    if (physicalTenantId == null) {
+      return supplier;
+    }
+    return () -> {
+      final String previous = getPropagatedPhysicalTenant();
+      setPropagatedPhysicalTenant(physicalTenantId);
+      try {
+        return supplier.get();
+      } finally {
+        if (previous != null) {
+          setPropagatedPhysicalTenant(previous);
+        } else {
+          clearPropagatedPhysicalTenant();
+        }
+      }
+    };
   }
 
   private static String asString(final Object value) {

--- a/spring-utils/src/test/java/io/camunda/spring/utils/PhysicalTenantContextTest.java
+++ b/spring-utils/src/test/java/io/camunda/spring/utils/PhysicalTenantContextTest.java
@@ -12,6 +12,7 @@ import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 import io.camunda.configuration.api.physicaltenants.PhysicalTenantIds;
 import jakarta.servlet.http.HttpServletRequest;
+import java.util.function.Supplier;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Test;
 import org.springframework.mock.web.MockHttpServletRequest;
@@ -23,6 +24,14 @@ class PhysicalTenantContextTest {
   @AfterEach
   void clearRequestContext() {
     RequestContextHolder.resetRequestAttributes();
+    // guarantee the propagation ThreadLocal never bleeds into a later test, even on failure
+    PhysicalTenantContext.clearPropagatedPhysicalTenant();
+  }
+
+  private void bindRequestWithPhysicalTenant(final String physicalTenantId) {
+    final HttpServletRequest request = new MockHttpServletRequest();
+    PhysicalTenantContext.setPhysicalTenantId(request, physicalTenantId);
+    RequestContextHolder.setRequestAttributes(new ServletRequestAttributes(request));
   }
 
   @Test
@@ -113,5 +122,57 @@ class PhysicalTenantContextTest {
     // when / then
     assertThat(PhysicalTenantContext.current())
         .isNotEqualTo(PhysicalTenantIds.DEFAULT_PHYSICAL_TENANT_ID);
+  }
+
+  @Test
+  void shouldReturnSupplierUnchangedWhenNoTenantInEffect() {
+    // given no request scope and no propagated tenant bound (nothing to capture)
+    final Supplier<String> original = () -> "result";
+
+    // when
+    final Supplier<String> wrapped = PhysicalTenantContext.propagateCurrent(original);
+
+    // then the supplier is returned unchanged and nothing is propagated when it runs
+    assertThat(wrapped).isSameAs(original);
+    assertThat(wrapped.get()).isEqualTo("result");
+    assertThat(PhysicalTenantContext.getPropagatedPhysicalTenant()).isNull();
+  }
+
+  @Test
+  void shouldBindCapturedTenantAndClearWhenNoPriorPropagatedValue() {
+    // given a request scope bound at capture time, but no prior propagated tenant
+    bindRequestWithPhysicalTenant("tenant-a");
+    final Supplier<String> wrapped =
+        PhysicalTenantContext.propagateCurrent(PhysicalTenantContext::getPropagatedPhysicalTenant);
+
+    // simulate a deferred run after the request scope is gone (e.g. session serialisation)
+    RequestContextHolder.resetRequestAttributes();
+
+    // when
+    final String observedDuringRun = wrapped.get();
+
+    // then the captured tenant was propagated during the run and cleared afterwards
+    assertThat(observedDuringRun).isEqualTo("tenant-a");
+    assertThat(PhysicalTenantContext.getPropagatedPhysicalTenant()).isNull();
+  }
+
+  @Test
+  void shouldRestorePriorPropagatedValueAfterRun() {
+    // given an already-propagated tenant on the current thread
+    PhysicalTenantContext.setPropagatedPhysicalTenant("outer");
+    // and a request scope bound at capture time so a different tenant is captured
+    bindRequestWithPhysicalTenant("inner");
+    final Supplier<String> wrapped =
+        PhysicalTenantContext.propagateCurrent(PhysicalTenantContext::getPropagatedPhysicalTenant);
+
+    // simulate a deferred run after the request scope is gone
+    RequestContextHolder.resetRequestAttributes();
+
+    // when
+    final String observedDuringRun = wrapped.get();
+
+    // then the captured tenant was propagated during the run and the prior value restored after
+    assertThat(observedDuringRun).isEqualTo("inner");
+    assertThat(PhysicalTenantContext.getPropagatedPhysicalTenant()).isEqualTo("outer");
   }
 }

